### PR TITLE
refactor(memo): incremental cycles

### PR DESCRIPTION
**Pull request by rgrinberg**
Improve performance by removing interruptible folds

 I doubt this impact anything performance wise, but this code is both shorter and clearer IMO. It's quite weird that the fold returns `Continue` at the end even though there's nothing to continue.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

**Comment by rgrinberg** [comment]
The benchmarks confirm that the minor collections have been reduced from 917 to 902. Minor collections are super fast, so this isn't really noteworthy, but it does confirm that the PR does something.

File: vendor/incremental-cycles/src/incremental_cycles.ml [hunk 1]
```diff
@@ -78,29 +54,24 @@ let rec visit_backward
...
+    visit_backward_aux g target mark vertex stack fuel (get_incoming g vertex)
+and visit_backward_aux g target mark vertex stack fuel = function
+  | [] -> visit_backward g target mark fuel stack
+  | y :: ys ->
+    if fuel = 0 then
+      (* There is no fuel left *)
+      VisitBackwardInterrupted
+    else if is_marked g y mark then
+      (* This vertex has already been visited, skip it *)
+      visit_backward_aux g target mark vertex stack fuel ys
```
Comment (ejgallego) [inline vendor/incremental-cycles/src/incremental_cycles.ml:66]: Why is not `fuel` decreased here anymore?

Comment (rgrinberg) [reply vendor/incremental-cycles/src/incremental_cycles.ml:66]: indeed it should be.

**Comment by rgrinberg** [comment]
Wasn't aware this library was extracted from Coq. I guess there's no point in modifying the code manually then.

**Comment by Alizter** [comment]
Do we have the original Coq code? If we can't reproduce the extraction then it seems kind of hard to maintain?

**Comment by snowleopard** [comment]
> Do we have the original Coq code? If we can't reproduce the extraction then it seems kind of hard to maintain?

Yep, the true sources are in https://gitlab.inria.fr/agueneau/incremental-cycles. Maybe we should vendor them too? With Dune's support for Coq, perhaps we can even regenerate things easily :)

**Comment by Alizter** [comment]
@snowleopard Thanks for the pointer! I think bringing Coq into the main Dune build would cause more complexity than good, I can't think of a non-intrusive way to do it. If @rgrinberg thinks that the perf increase here is something we want, then perhaps we can try to reproduce it in the Coq code. Not something I want to do anytime soon however.

**Comment by snowleopard** [comment]
@ejgallego I'd love to have some support for (on-demand) re-extraction of code from Coq!

**Pull request closed without merging**